src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -515,6 +516,7 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
         // primitives.
         resolve_primitive(&path_root, TypeNS)
             .or_else(|| self.resolve_path(&path_root, TypeNS, module_id))
+            .or_else(|| self_id.map(|id| Res::Def(self.cx.tcx.def_kind(id), id)))
```

Comment:
This seems like the wrong fix; it will use other items in the same scope if they happen to have the same name.  This should come first and only then resolve_primitive and resolve_path. This is https://github.com/rust-lang/rust/issues/84827#issuecomment-830840878; please add it as a test case.